streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java

```diff
@@ -450,6 +456,18 @@ public void advanceWallClockTime(final long advanceMs) {
         final V value = valueDeserializer.deserialize(record.topic(), record.value());
         return new ProducerRecord<>(record.topic(), record.partition(), record.timestamp(), key, value);
     }
+    
+    /**
+     * Records sent with this {@link Producer} are streamed to the topology.
+     * <p>
+     * This is useful when testing processes that are also using "normal" kafka producer, tipically in
+     * a different thread. 
+     * 
+     * @return the producer.
+     */
+    public final Producer<byte[], byte[]> getProducer() {
```

Comment:
This is a meta comment: could you elaborate a bit more on use cases where people want to use this producer.send than calling `pipeInput` directly? To me they seems to be covering the same pattern. More specifically could you give me an example topology where testing it would require using a producer than piping data directly to the source topic?

Reply:
I agree with Guozhang. Don't really understand the purpose of this.

Also note: this would be a public API change and requires a KIP -- if want to add this, we should have a separate JIRA and not add this change to this bug fix PR.

Reply:
Still not clear why you cannot use `pipeInput()` -- I guess the correct way would be to implement a custom `Producer` that takes the `TopologyTestDriver` as parameter. You pass with `Producer` to your "watchdog" and internally in `.send()` you call `pipeInput()`. Let us know if this would work for you.

I would prefer to not expose the producer in the first place. Also, if we want to expose it, we need a KIP.

Reply:
Yes I can do that way but that's exactly what the MockProducer does, so, since I'm a lazy programmer :) and since I think it could be useful to others, I would like to reuse it.
Anyway I can understand your reluctance in exposing the (Mock)Producer, so do you think we can just expose it as a protected instance variable or it's better to remove it (also from tests) for a better future overall solution?
In any case I think that a solution that let's us able to test mixed scenarios, I mean kafka-streams topology used alongside a producer, should be provided.